base/pkg/libgit2.jl

```diff
@@ -0,0 +1,438 @@
+                force=force,
+                msg="pkg.libgit2.branch: moving to $branch_name")
+        finally
+            finalize(cmt)
+        end
+    end
+    try
+        if !isempty(track) # setup tracking #TODO: what is branch tracks other then "origin" remote
+            try
+                with(GitConfig, repo) do cfg
+                    set!(cfg, "branch.$branch_name.remote", GitConst.REMOTE_ORIGIN)
+                    set!(cfg, "branch.$branch_name.merge", name(branch_ref))
+                end
+            catch
+                warn("Please provide remote tracking for branch '$branch_name' in '$(path(repo))'")
+            end
+        end
+
+        # switch head to the branch
+        set_head && head!(repo, branch_ref)
+    finally
+        finalize(branch_ref)
+    end
+    return
+end
+
+""" git checkout [-f] --detach <commit> """
+function checkout!(repo::GitRepo, commit::AbstractString = "";
+                  force::Bool = true)
+    # nothing to do
+    isempty(commit) && return
+
+    # grab head name
+    head_name = GitConst.HEAD_FILE
+    try
+        with(head(repo)) do head_ref
+            head_name = shortname(head_ref)
+            # if it is HEAD use short OID instead
+            if head_name == GitConst.HEAD_FILE
+                head_name = string(Oid(head_ref))
+            end
+        end
+    end
+
+    # search for commit to get a commit object
+    obj = get(GitAnyObject, repo, Oid(commit))
+    obj == nothing && return
+    try
+        peeled = peel(obj, GitConst.OBJ_COMMIT)
+        peeled == nothing && return
+        opts = force ? CheckoutOptionsStruct(checkout_strategy = GitConst.CHECKOUT_FORCE) :
+                           CheckoutOptionsStruct()
+        try
+            # detach commit
+            obj_oid = Oid(peeled)
+            ref = create_reference(repo, obj_oid,
+                force=force,
+                msg="pkg.libgit2.checkout: moving from $head_name to $(string(obj_oid))")
+            finalize(ref)
+
+            # checkout commit
+            checkout_tree(repo, peeled, options = opts)
+        finally
+            finalize(peeled)
+        end
+    finally
+        finalize(obj)
+    end
+end
+
+""" git clone [-b <branch>] [--bare] <url> <dir> """
+function clone(url::AbstractString, path::AbstractString;
+               branch::AbstractString="",
+               bare::Bool = false,
+               remote_cb::Ptr{Void} = C_NULL)
+    # setup colne options
+    clone_opts = CloneOptionsStruct(
+                    bare = Int32(bare),
+                    checkout_branch = isempty(branch) ? Ptr{UInt8}(C_NULL) : pointer(branch),
+                    remote_cb = remote_cb
+                )
+
+    # start cloning
+    clone_opts_ref = Ref(clone_opts)
+    repo_ptr_ptr = Ref{Ptr{Void}}(C_NULL)
+    @check ccall((:git_clone, :libgit2), Cint,
+            (Ptr{Ptr{Void}}, Ptr{UInt8}, Ptr{UInt8}, Ref{CloneOptionsStruct}),
```

Comment:
Here and elsewhere, you should use `Cstring`, not `Ptr{UInt8}`, when passing Julia strings as arguments to `ccall` where a NUL-terminated UTF-8 string is expected.